File: src/endpoints/nest/get_index.py
from . import __types__ as types
from src import output
from src.libs.db import DB, Table
from src.libs.object import json_parse
from src.modules.verify import checking_token
from src.modules.mod import MOD
from ..app import __libs__ as app_libs
from ..article import __libs__ as article_libs

async def get_index(params: dict = {}, req = None, _db: DB = None, _check_token = True):

    # set values
    result = None
    db = _db if _db else DB().connect()

    try:
        # set params
        params = types.GetIndex(**params)

        # checking token
        if _check_token: checking_token(req, db)

        # set fields
        fields = params.fields.split(',') if params.fields else None

        # set mod
        mod = MOD(params.mod or '')

        # set where
        where = []
        if params.app_srl:
            where.append(f'and app_srl = {params.app_srl}')
        if params.code:
            where.append(f'and code LIKE "{params.code}"')
        if params.name:
            where.append(f'and name LIKE "%{params.name}%"')

        # get total
        total = db.get_count(
            table_name = Table.NEST.value,
            where = where,
        )
        if not (total > 0): raise Exception('No data', 204)

        # get index
        index = db.get_items(
            table_name = Table.NEST.value,
            fields = fields,
            where = where,
            limit = {
                'size': params.size,
                'page': params.page,
            },
            order = {
                'order': params.order,
                'sort': params.sort,
            },
            unlimited = params.unlimited,
        )

        # transform items
        def transform_item(item: dict) -> dict:
            if 'json' in item and item['json']:
                item['json'] = json_parse(item['json'])
            # MOD / app
            if mod.check('app') and item.get('app_srl'):
                item['app'] = app_libs.get_item(
                    _db = db,
                    srl = item.get('app_srl'),
                    fields = [ 'srl', 'code', 'name' ],
                )
            # MOD / count-article
            if mod.check('count-article'):
                item['count_article'] = article_libs.get_count(
                    _db = db,
                    where = [ f'nest_srl = {item['srl']}' ],
                )
            return item
        index = [ transform_item(item) for item in index ]

        # set result
        result = output.success({
            'message': 'Complete get nest index.',
            'data': {
                'total': total,
                'index': index,
            },
        })
    except Exception as e:
        result = output.exc(e)
    finally:
        if not _db and db: db.disconnect()
        return result

File: src/routes/nest/get-index.ts
import DB, { db } from '@/classes/DB'
import ServiceError from '@/classes/ServiceError'
import Service from '@/classes/Service'
import MOD from '@/classes/MOD'
import { parseJSON } from '@/libs/objects'
import * as articleHelper from '@/routes/article/__helper'
import type { NestModel } from './__model'

type GetIndexParams = {
  query: NestModel['getIndexQuery']
  service: Service
}

export default async function getIndex({ query, service }: GetIndexParams)
{
  try
  {
    // set assets
    let _table = `${DB.TABLE.NEST} AS n`
    let _where: string[] = []
    let _values: ZZ = {}
    let _join: string[] = []
    const _field = query.field ? query.field.split(',') : ''

    // set base params
    if (query.app)
    {
      _where.push(`AND app_srl = ${query.app}`)
    }
    if (query.code)
    {
      _where.push(`AND code LIKE \'${query.code}\'`)
    }
    if (query.name)
    {
      _where.push(`AND name LIKE \'%${query.name}%\'`)
    }

    // get count
    const count = db.getCount({
      table: _table,
      where: _where,
    })
    if (count.data <= 0) throw new ServiceError('No data', { status: 204 })

    // get index
    let index = db.getIndex({
      table: _table,
      field: _field,
      where: _where,
      join: _join,
      order: query.order,
      page: query.page ?? 1,
      size: query.size ?? service.preference['nest.index.size'],
      values: _values,
    })

    // set MOD
    const _mod: MOD = new MOD(query.mod)

    // transform index
    index.data = index.data.map((o: ZZ) => {
      // MOD / app
      if (_mod.check('app') && o.app_srl)
      {
        o.app = db.getData({
          table: DB.TABLE.APP,
          field: 'srl,code,name',
          where: `srl = ${o.app_srl}`,
        }).data
      }
      // MOD / count-article
      if (_mod.check('count-article'))
      {
        o.count_article = articleHelper.count({
          where: `nest_srl = ${o.srl}`,
        })
      }
      return {
        ...o,
        json: parseJSON(o.json),
      }
    })

    return {
      total: count.data,
      index: index.data,
    }
  }
  catch (_e: any)
  {
    throw new ServiceError('Failed to get Nest index.', {
      status: _e.status,
      text: _e.message,
      cause: _e,
    })
  }
}
